src/uct/tcp/tcp_sockcm_ep.c

```diff
@@ -24,10 +24,298 @@ ucs_status_t uct_tcp_sockcm_ep_disconnect(uct_ep_h ep, unsigned flags)
+    if (priv_data_ret < 0) {
+        ucs_assert(priv_data_ret > UCS_ERR_LAST);
+        status = (ucs_status_t)priv_data_ret;
+        ucs_error("tcp_sockcm private data pack function failed with error: %s",
+                  ucs_status_string(status));
+        goto out;
+    } else if (priv_data_ret > (uct_tcp_sockcm_ep_get_cm(cep)->priv_data_len -
+                                sizeof(uct_tcp_sockcm_priv_data_hdr_t))) {
+        status = UCS_ERR_EXCEEDS_LIMIT;
+        ucs_error("tcp_sockcm private data pack function returned %zd "
+                  "(max: %zu)", priv_data_ret,
+                  uct_tcp_sockcm_ep_get_cm(cep)->priv_data_len -
+                  sizeof(uct_tcp_sockcm_priv_data_hdr_t));
+        goto out;
+    }
+
+    hdr->length          = priv_data_ret;
+    cep->comm_ctx.length = sizeof(*hdr) + hdr->length;
+
+    status = uct_tcp_sockcm_ep_progress_send(cep, is_async);
+
+out:
+    return status;
+}
+
+static ucs_status_t uct_tcp_sockcm_ep_fd_to_remote_dev_addr(int fd,
+                                                            uct_device_addr_t **dev_addr_p,
+                                                            size_t *dev_addr_len_p)
+{
+    struct sockaddr_storage peer_addr = {0};
+    char peer_str[UCS_SOCKADDR_STRING_LEN];
+    size_t peer_addr_len;
+    ucs_status_t status;
+
+    status = ucs_socket_getpeername(fd, &peer_addr);
+    if (status != UCS_OK) {
+        return status;
+    }
+
+    status = ucs_sockaddr_sizeof((struct sockaddr*)&peer_addr, &peer_addr_len);
+    if (status != UCS_OK) {
+        return status;
+    }
+
+    *dev_addr_p = ucs_malloc(peer_addr_len, "IB device address");
+    if (*dev_addr_p == NULL) {
+        ucs_error("failed to allocate device address");
+        return UCS_ERR_NO_MEMORY;
+    }
+
+    ucs_debug("fd %d: peer_addr: %s of length: %zu", fd,
+              ucs_sockaddr_str((const struct sockaddr*)&peer_addr, peer_str,
+                               UCS_SOCKADDR_STRING_LEN), peer_addr_len);
+
+    memcpy(*dev_addr_p, &peer_addr, peer_addr_len);
+    *dev_addr_len_p = peer_addr_len;
+    return UCS_OK;
+}
+
+static ucs_status_t uct_tcp_sockcm_ep_server_invoke_conn_req_cb(uct_tcp_sockcm_ep_t *cep)
+{
+    uct_tcp_sockcm_priv_data_hdr_t *hdr = (uct_tcp_sockcm_priv_data_hdr_t *)
+                                          cep->comm_ctx.buf;
+    char                 ifname_str[UCT_DEVICE_NAME_MAX];
+    uct_cm_remote_data_t remote_data;
+    uct_device_addr_t    *dev_addr;
+    size_t               addr_length;
+    ucs_status_t         status;
+
+    /* get the local interface name associated with the connected fd */
+    status = ucs_sockaddr_get_ifname(cep->fd, ifname_str, UCT_DEVICE_NAME_MAX);
+    if (UCS_OK != status) {
+        return status;
+    }
+
+    /* get the device address of the remote peer associated with the connected fd */
+    status = uct_tcp_sockcm_ep_fd_to_remote_dev_addr(cep->fd, &dev_addr, &addr_length);
```

Comment:
why not just define dev_addr as sockaddr_storage and call ucs_socket_getpeername directly here?
ucs_socket_getpeername could also return the address length